### ISSUE TITLE
pass a URL with jimp.read and have option to pass "rejectUnauthorized : false" to the request

### ISSUE THREAD

**Issue by bastoune** [oliver-moran/jimp#609]
# Expected Behavior
I would like to pass an option `rejectUnauthorized : false` to the Jimp.read()` method when I pass a URL to it.
I want to read the image even if there is a problem with the certificate.
I don't want to configure the behaviour globally to the server.

# Current Behavior
If I have somehow troubles to get the certificate of the URL, I get the following error => Error: unable to verify the first certificate

# Failure Information (for bugs)

> Error
> unable to verify the first certificate
> _tls_wrap.js in TLSSocket.<anonymous> at line 1105:38
> events.js in emitNone at line 106:13
> events.js in TLSSocket.emit at line 208:7
> _tls_wrap.js in TLSSocket._finishInit at line 639:8
> _tls_wrap.js in TLSWrap.ssl.onhandshakedone at line 469:38

## Steps to Reproduce
Jimp.read('https://404store.com/2017/12/08/Random-random-30798866-500-325.jpg')

## Context
- Jimp Version: 0.4.0
- Operating System: Unbuntu Server 16.04
- Node version: v8.11.1

## Failure Logs

> Error
> unable to verify the first certificate
> _tls_wrap.js in TLSSocket.<anonymous> at line 1105:38
> events.js in emitNone at line 106:13
> events.js in TLSSocket.emit at line 208:7
> _tls_wrap.js in TLSSocket._finishInit at line 639:8
> _tls_wrap.js in TLSWrap.ssl.onhandshakedone at line 469:38

**Comment by hipstersmoothie** [comment]
related to #267

**Comment by turakvlad** [comment]
@hipstersmoothie, hey. 

I had the same issue. Do you have any workaround on it except for [this one](https://github.com/oliver-moran/jimp/issues/267#issuecomment-360546370)?

**Comment by hipstersmoothie** [comment]
I feel like you could just do a custom constructor in the mean time.  I just release d https://github.com/oliver-moran/jimp/pull/628. In v0.5.1 you can do the following. This allows you to load your URL however you want and still fallback to jimp image parsing.

```js
const Jimp = require('jimp');
const phin = require('phin');

Jimp.appendConstructorOption('Custom Url', options => options.url, function(
  resolve,
  reject,
  options
) {
  phin(options, (err, res) => {
    if (err) {
      return reject(err);
    }

    this.parseBitmap(res.body, null, err => {
      if (err) {
        return reject(err);
      }

      resolve();
    });
  });
});

test();

async function test() {
  const image = await Jimp.read({
    url: 'http://pngimg.com/uploads/cat/cat_PNG50539.png'
  });

  image.writeAsync('test.png');
}
```

Is this sufficient? @bastoune @turakvlad 

I'm also looking to exposing more options to the request, but the library [we use](https://www.npmjs.com/package/phin) might not have to many customization options.

**Comment by hipstersmoothie** [comment]
@bastoune @turakvlad v0.5.2 has a new constructor option that allows you to pass the headers. Can you try it out for me? Does this suit your usecase?

```js
await Jimp.read({ url: '...', headers: { ... } })
```

**Comment by bastoune** [comment]
Hi @hipstersmoothie , thanks for replies.
 I am sorry I had to find another way and I am now running this part of code in python. 

I will answer you here if I change my mind ;)

**Comment by turakvlad** [comment]
@hipstersmoothie, hey.

Thanks for helping us with this issue. Please, have a look at this [Runkit](https://runkit.com/turakvlad/5baa1c31cd3e740012900bc6). To be honest, I am not sure that the `headers` will help us. What should I pass in the `headers` to ignore HTTPS errors?

For example, Node.js has [`rejectUnauthorized`](https://nodejs.org/api/https.html#https_https_request_url_options_callback) option for this. The request library has [`strictSSL`](https://www.npmjs.com/package/request#requestoptions-callback) option (but I'm sure that it uses Node.js's `rejectUnauthorized` under the hood). Puppeteer uses [`ignoreHTTPSErrors`](https://github.com/GoogleChrome/puppeteer/blob/v1.8.0/docs/api.md#puppeteerconnectoptions).

`rejectUnauthorized` is not a header option, it's a [TLS](https://nodejs.org/api/tls.html) option.

If I can help you in anyway, please just let me know.

**Comment by turakvlad** [comment]
@hipstersmoothie, hi!

I've spent some time to examine this issue. [This](https://runkit.com/turakvlad/5bae50cb4a0edb00114e003f) works great. It's possible to pass `rejectUnauthorized` thankfully to [this](https://github.com/turakvlad/phin/blob/master/lib/phin.js#L59) line of code in phin. And [later](https://github.com/turakvlad/phin/blob/master/lib/phin.js#L124) in the code the `https.request()` method will be executed with the options we need.

Actually, we can pass all the options that [`http.request(...)`](https://nodejs.org/api/http.html#http_http_request_options_callback) and [`https.request(...)`](https://nodejs.org/api/https.html#https_https_request_options_callback) modules support:

```javascript
await Jimp.read({ url: '...', headers: { ... } })
```

It's a little bit hard to understand without looking at the [`phin`](https://github.com/ethanent/phin)'s code base.

If you want, I can document this via PR. Just tell me what `README.md` I should change? Maybe [this one](https://github.com/oliver-moran/jimp/blob/master/packages/jimp/README.md)?

I think you can close this issue.

Thanks.

**Comment by turakvlad** [comment]
@hipstersmoothie, thanks! I am glad to help.

I will make a PR tomorrow.

**Comment by turakvlad** [comment]
@hipstersmoothie, I am sorry for a delay. Will make a PR tomorrow.

**Comment by hipstersmoothie** [comment]
@turakvlad no worries! take your time

**Comment by hipstersmoothie** [comment]
closed via #637
